aggregators/converter.go

```diff
@@ -30,6 +30,277 @@ const (
+	partitionedMetricsBuilderPool.Put(p)
+}
+
+func (p *partitionedMetricsBuilder) processEvent(e *modelpb.APMEvent) {
+	switch e.Type() {
+	case modelpb.TransactionEventType:
+		repCount := e.GetTransaction().GetRepresentativeCount()
+		if repCount <= 0 {
+			// BUG we should add a service summary metric
+			return
+		}
+		duration := e.GetEvent().GetDuration().AsDuration()
+		p.addTransactionMetrics(e, repCount, duration)
+		p.addServiceTransactionMetrics(e, repCount, duration)
+		for _, dss := range e.GetTransaction().GetDroppedSpansStats() {
+			p.addDroppedSpanStatsMetrics(dss, repCount)
+		}
+	case modelpb.SpanEventType:
+		target := e.GetService().GetTarget()
+		repCount := e.GetSpan().GetRepresentativeCount()
+		destSvc := e.GetSpan().GetDestinationService().GetResource()
+		if repCount <= 0 || (target == nil && destSvc == "") {
+			// BUG we should add a service summary metric
+			return
+		}
+		p.addSpanMetrics(e, repCount)
+	default:
+		// All other event types should add an empty service metrics,
+		// for adding to service summary metrics.
+		p.addServiceSummaryMetrics()
+	}
+}
+
+func (p *partitionedMetricsBuilder) addTransactionMetrics(e *modelpb.APMEvent, count float64, duration time.Duration) {
+	var key aggregationpb.TransactionAggregationKey
+	setTransactionKey(e, &key)
+
+	mb := p.get(p.hasher.Chain(&key))
+	mb.transactionAggregationKey = key
+
+	hdr := hdrhistogram.New()
+	hdr.RecordDuration(duration, count)
+	setHistogramProto(hdr, &mb.transactionHistogram)
+	mb.transactionMetrics.Histogram = &mb.transactionHistogram
+	mb.keyedTransactionMetricsSlice = mb.keyedTransactionMetricsArray[:]
+}
+
+func (p *partitionedMetricsBuilder) addServiceTransactionMetrics(e *modelpb.APMEvent, count float64, duration time.Duration) {
+	var key aggregationpb.ServiceTransactionAggregationKey
+	setServiceTransactionKey(e, &key)
+
+	mb := p.get(p.hasher.Chain(&key))
+	mb.serviceTransactionAggregationKey = key
+
+	if mb.transactionMetrics.Histogram == nil {
+		// mb.TransactionMetrics.Histogram will be set if the event's
+		// transaction metric ended up in the same partition.
+		hdr := hdrhistogram.New()
+		hdr.RecordDuration(duration, count)
+		setHistogramProto(hdr, &mb.transactionHistogram)
+	}
+	mb.serviceTransactionMetrics.Histogram = &mb.transactionHistogram
+	switch e.GetEvent().GetOutcome() {
+	case "failure":
+		mb.serviceTransactionMetrics.SuccessCount = 0
+		mb.serviceTransactionMetrics.FailureCount = count
+	case "success":
+		mb.serviceTransactionMetrics.SuccessCount = count
+		mb.serviceTransactionMetrics.FailureCount = 0
+	}
+	mb.keyedServiceTransactionMetricsSlice = mb.keyedServiceTransactionMetricsArray[:]
+}
+
+func (p *partitionedMetricsBuilder) addDroppedSpanStatsMetrics(dss *modelpb.DroppedSpanStats, repCount float64) {
+	var key aggregationpb.SpanAggregationKey
+	setDroppedSpanStatsKey(dss, &key)
+
+	mb := p.get(p.hasher.Chain(&key))
+	i := len(mb.keyedSpanMetricsSlice)
+	if i == 128 { //len(mb.keyedSpanMetrics) {
```

Comment:
```suggestion
	if i == 128 {
```

Reply:
How about I change it to `if i == len(mb.keyedSpanMetricsArray) {`? That way it's a bit more self-documenting.

Reply:
going with that, can follow up if needed

Reply:
Sounds good. I was a bit conflicted between the two but didn't have a strong preference either way.